### PR TITLE
Fix xnu_wait stopping prematurely (#7744)

### PR DESCRIPTION
**Pull request by puxlit**
The scenarios mentioned in #7744 now behave in line with expectations.

```
$ cp /bin/sleep /tmp/sleep # SIP shenanigans
$ ~/bin/r2 -d /tmp/sleep
= attach 45562 45562
bin.baddr 0x100000000
Using 0x100000000
asm.bits 64
 -- Charlie! We are here.
[0x100006000]> ?V;e dbg.
1.6.0-git aka 1.5.0-120-gb41c5a30f commit 15056
dbg.aftersyscall = true
dbg.args =
dbg.backend = native
dbg.bep = loader
dbg.bpinmaps = true
dbg.bpsize = 1
dbg.btalgo = fuzzy
dbg.btdepth = 128
dbg.clone = false
dbg.consbreak = false
dbg.execs = false
dbg.exitkills = true
dbg.follow = 64
dbg.follow.child = false
dbg.forks = false
dbg.hwbp = 0
dbg.libs =
dbg.profile =
dbg.slow = false
dbg.status = false
dbg.swstep = false
dbg.threads = false
dbg.trace = false
dbg.trace.inrange = false
dbg.trace.libs = true
dbg.trace.tag = 0
dbg.unlibs =
[0x100006000]> dc
usage: sleep seconds

==> Process finished

[0x00000000]> ood 10
= attach 45562 45562
File dbg:///private/tmp/sleep  10 reopened in read-write mode
= attach 45565 45565
= attach 45565 45562
= attach 45565 45565
[0x100006000]> dc

==> Process finished

[0x00000000]>
```

```
$ echo 'int main(void){while(1);}' | gcc -Wall -Werror -O0 -xc -o ./busyloop -
$ ~/bin/r2 -d ./busyloop
= attach 45600 45600
bin.baddr 0x100000000
Using 0x100000000
asm.bits 64
 -- You haxor! Me jane?
[0x100003000]> dc
^Cmessage interrupted
[0x00000000]>
```

For the case where we attach to an existing process, it's worth noting that (currently) the inferior isn't suspended. In line with that behaviour, a `dc` + `^C` will yield the prompt but won't suspend the inferior. (Perhaps this should change; perhaps the inferior should be suspended upon attaching, as well as `dc` + `^C`.)

### CODE REVIEW

**Comment by radare** [comment]
Theres a place where this ^C is handled to send a signal to the child, maybe this is how we should do it for XNU too. This was done iirc for windows because processes are never a child of the debugger

**Comment by radare** [comment]
Ping

**Comment by puxlit** [comment]
Sounds fair. I'll see how `gdb` and `lldb` behave.

**Comment by XVilka** [comment]
@puxlit any updates?

**Comment by radare** [comment]
ping

**Comment by puxlit** [comment]
Signal handling appears non-existent, so I've elected to stop the inferior with Mach calls instead.

**Comment by radare** [comment]
Sounds good

File: libr/debug/p/native/xnu/xnu_debug.c [hunk 1]
```diff
@@ -198,6 +199,47 @@ int xnu_detach(RDebug *dbg, int pid) {
...
+#else
+	kern_return_t kr;
+	task_t task;
+	int suspend_count;
+
+	task = pid_to_task (pid);
+	if (!task) return false;
+
+	suspend_count = task_suspend_count (task);
+	if (suspend_count == -1) return false;
```
Comment (radare) [inline libr/debug/p/native/xnu/xnu_debug.c]: Missing braces

File: libr/debug/p/native/xnu/xnu_debug.c [hunk 2]
```diff
@@ -198,6 +199,47 @@ int xnu_detach(RDebug *dbg, int pid) {
...
+#if XNU_USE_PTRACE
+	eprintf ("xnu_stop: not implemented\n");
+	return false;
+#else
+	kern_return_t kr;
+	task_t task;
+	int suspend_count;
+
+	task = pid_to_task (pid);
+	if (!task) return false;
```
Comment (radare) [inline libr/debug/p/native/xnu/xnu_debug.c]: Missing braces

File: libr/debug/p/native/xnu/xnu_debug.c [hunk 3]
```diff
@@ -198,6 +199,47 @@ int xnu_detach(RDebug *dbg, int pid) {
...
+	if (suspend_count > 1) return false;  // This is unexpected.
+
+	kr = task_suspend (task);
+	if (kr != KERN_SUCCESS) {
+		eprintf ("failed to suspend task\n");
+		return false;
+	}
+
+	suspend_count = task_suspend_count (task);
+	if (suspend_count != 1) return false;  // This is unexpected.
```
Comment (radare) [inline libr/debug/p/native/xnu/xnu_debug.c]: Missing braces

**Comment by radare** [comment]
Lgtm. Fix the comments and ill merge

Thanks!!

File: libr/debug/p/native/xnu/xnu_debug.c (hunk 2 above)
Comment (puxlit) [inline libr/debug/p/native/xnu/xnu_debug.c]: [Style guide](https://github.com/radare/radare2/blame/266686278af37340c0099dc95384a8544dc0413b/CONTRIBUTING.md#L46) seems to permit braceless ifs at the moment (per lines 46, 59–66, and 97–98); might require updating. 😛

**Pull request merged**